File: Source/Runtime/CPlusplus/Graphics/Bibim/Texture2D.cpp
#include <Bibim/PCH.h>
#include <Bibim/Texture2D.h>
#include <Bibim/GraphicsDevice.h>
#include <Bibim/CheckedRelease.h>

namespace Bibim
{
    Texture2D::Texture2D(GraphicsDevice* graphicsDevice)
        : graphicsDevice(graphicsDevice),
          width(0),
          height(0),
          surfaceWidth(0),
          surfaceHeight(0),
          d3dTexture(nullptr)
    {
    }

    Texture2D::Texture2D(GraphicsDevice* graphicsDevice, int width, int height, int surfaceWidth, int surfaceHeight)
        : graphicsDevice(graphicsDevice),
          width(width),
          height(height),
          surfaceWidth(surfaceWidth),
          surfaceHeight(surfaceHeight),
          d3dTexture(nullptr)
    {
    }

    Texture2D::~Texture2D()
    {
        CheckedRelease(d3dTexture);
    }

    void Texture2D::Setup(IDirect3DTexture9* d3dTexture, int width, int height, int surfaceWidth, int surfaceHeight)
    {
        CheckedRelease(this->d3dTexture);
        this->d3dTexture = d3dTexture;
        this->width = width;
        this->height = height;
        this->surfaceWidth = surfaceWidth;
        this->surfaceHeight = surfaceHeight;
    }
}

File: Source/Runtime/CSharp/Halak.Bibim.Runtime/Graphics/Texture2D.cs
using System;
using System.Collections.Generic;
using System.Text;
using Halak.Bibim.Asset;

namespace Halak.Bibim.Graphics
{
    public abstract class Texture2D : GameAsset
    {
        #region Fields
        private GraphicsDevice graphicsDevice;
        #endregion

        #region Properties
        public GraphicsDevice GraphicsDevice
        {
            get { return graphicsDevice; }
            set
            {
                if (graphicsDevice != value)
                {
                    graphicsDevice = value;
                }
            }
        }

        public int Width
        {
            get;
            private set;
        }

        public int Height
        {
            get;
            private set;
        }

        public int SurfaceWidth
        {
            get;
            private set;
        }

        public int SurfaceHeight
        {
            get;
            private set;
        }
        #endregion

        #region Constructors
        public Texture2D()
        {
        }

        public Texture2D(GraphicsDevice graphicsDevice)
        {
            this.graphicsDevice = graphicsDevice;
        }
        #endregion

        #region Methods
        protected void Setup(int width, int height, int surfaceWidth, int surfaceHeight)
        {
            Width = width;
            Height = height;
            SurfaceWidth = surfaceWidth;
            SurfaceHeight = surfaceHeight;
        }
        #endregion
    }
}
